pkg/policy/selectorcache.go

```diff
@@ -0,0 +1,335 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"sort"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+// CachedSelector represents an identity selector owned by the selector cache
+type CachedSelector interface {
+	// GetSelections returns the cached set of numeric identities
+	// selected by the CachedSelector.  The retuned slice must NOT
+	// be modified, as it is shared among multiple users.
+	GetSelections() []identity.NumericIdentity
+
+	// String returns the string representation of this selector.
+	// Used as a map key.
+	String() string
+}
+
+// CachedSelectionUser inserts selectors into the cache and gets update
+// callbacks whenever the set of selected numeric identities change for
+// the CachedSelectors pushed by it.
+type CachedSelectionUser interface {
+	// IdentitySelectionUpdated implementations MUST NOT call back
+	// to selector cache while executing this function!
+	IdentitySelectionUpdated(selector CachedSelector, selections, added, deleted []identity.NumericIdentity)
+}
+
+// IdentitySelector is the internal interface for all selectors in the
+// selector cache.
+//
+// IdentitySelector represents the mapping of an EndpointSelector
+// to a slice of identities. These mappings are updated via two
+// different processes:
+//
+// 1. When policy rules are changed these are added and/or deleted
+// depending on what selectors the rules contain. Cached selections of
+// new IdentitySelectors are pre-populated from the set of currently
+// known identities.
+//
+// 2. When reachacble identities appear or disappear, either via local
+// allocation (CIDRs), or via the KV-store (remote endpoints). In this
+// case all existing IdentitySelectors are walked through and their
+// cached selections are updated as necessary.
+//
+// In both of the above cases the set of existing IdentitySelectors is
+// write locked.
+//
+// To minimize the upkeep the identity selectors are shared accross
+// all IdentityPolicies, so that only one copy exists for each
+// IdentitySelector. Users of the SelectorCache take care of creating
+// IdentitySelectors as needed by identity policies. The set of
+// IdentitySelectors is read locked during an IdentityPolicy update so
+// that the the policy is always updated using a coherent set of
+// cached selections.
+//
+// IdentitySelector is used as a map key, so it must not be implemented by a
+// map, slice, or a func, or a runtime panic will be triggered. In all
+// cases below IdentitySelector is being implemented by structs.
+type IdentitySelector interface {
```

Comment:
By the way I think this can be hidden, identitySelector since the interface isn't used outside this package.

Reply:
Will do for the next round.

Reply:
Done!